### DISCUSSION TITLE
Modify IL without changing machine code

### DISCUSSION THREAD

**Discussion by Ephemerua** [Vector35/binaryninja-api#3096] [Ideas] [answered]
Binja's IL is pretty cool, but I found currently the IL cannot be modified, it could only reflect assembly's change.
To do analysis such as deobfuscation, I think it makes sense to be able to modify LLIL or MLIL directly, because it is very cumbersome to directly modify the asm, and different architectures need different patch ways. Besides, most of the time I just want neat pseudo C code rather than a clean binary file.

Does/Will binja provide a ILBuilder API like LLVM's IRBuilder that allows users to construct IL directly?

**Comment by brinlyau** [comment, marked as answer]
Not a BINJA dev but: can’t you just do an analysis pass? (Also there’s a
C-like pseudo codeview now)

**Reply by psifertex** [reply]
Yup! Analysis passes (or workflows as they are called now) enable you to modify the IL and are designed for exactly such a use case as this: 

https://github.com/Vector35/binaryninja-api/tree/dev/examples/workflows

**Reply by Ephemerua** [reply]
Wow thanks!! I will look into that.

**Reply by brinlyau** [reply]
you should probably not mark my answer as “selected” - mark jordans one until I expand with more info  but that may be a couple days bc I’m on mobile :)

**Reply by psifertex** [reply]
It's the thread that gets selected. You were spot on, I intentionally replied instead of starting a new thread 😉
